src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Interlocked.cs

```diff
@@ -23,6 +23,25 @@ public static int CompareExchange(ref int location1, int value, int comparand)
 #endif
         }
 
+        // This is used internally by NativeAOT runtime in cases where having a managed
+        // ref to the location is unsafe (Ex: it is the syncblock of a pinned object).
+        // The intrinsic expansion for this overload is exactly the same as for the `ref int`
+        // variant and will go on the same path since expansion is triggered by the name and
+        // return type of the method.
+        // The important part is avoiding `ref *location` in the unexpanded scenario, like
+        // in a case when compiling the Debug flavor of the app.
+        [Intrinsic]
+        internal static unsafe int CompareExchange(int* location1, int value, int comparand)
```

Comment:
I also wonder why some of CompareExchange overloads are AggressiveInlining and some are not? Is that intentional or just does not matter?

Reply:
I think it is needed for the ones that contain non-trivial amount of code. For the trivial ones, it is probably just copy&paste.

Reply:
So I will keep the new one without AgressiveInlining and existing ones unchanged. 
It seems like in perf-interesting scenarios this will not matter either way.